### DISCUSSION TITLE
Support for max() in objective

### DISCUSSION THREAD

**Discussion by cahly** [do-mpc/do-mpc#148] [Ideas]
Hi

I am trying to create an optimization problem where i am interested in minimizing the maximum value over the horizon of some auxillary expression. Ideally i would want to add something like k*max(expr) to the objective function, where expr is the expression of which i try to minimize the maximum value, and k is the cost associated with that maximum value

I thought there was a workaround doing something like this.

mpc.set_nl_cons(expr_name='myExpr',expr=myExpr,ub=0,soft_constraint=True,penalty_term_cons=k)  

which i thought would add a **single** slack variable to the objective function multiplied by the cost k, which would have the same effect. However, it seems to add a slack variable for **each** time step, which is not what i want. I know the objective function becomes non-seperable if you only have a single variable, but it is not clear to me if that is hard requirement, or if it is just how mpc usually is formulated. 
 
It would be great to see the support for adding max() or even ininity norm to the objective function. Or the quicker but more obscure fix i guess would be a change to allow an option on the set_nl_cons() method that only adds a single slack variable, which applies to every time step, to the objective function.

Great work btw.
BR

**Comment by 4flixt** [comment]
HI Cahly, thank you for your question! I think your proposal to have an option for ``set_nl_cons`` to only introduce a single slack variable over the horizon is very good. It is not difficult to change in the backend and I will create an issue for this.
I think your workaround might work in that case.

Your other question regarding max or inifinty norm in the objective: We don't expect the optimizer to perform reasonably well in that case and don't plan to implement this feature in the future.

One last thing: I am moving this issue to the discussions page and will create a new issue regarding the slack variables in ``set_nl_cons``.

Best 
4flixt

**Reply by 4flixt** [reply]
Voila: #149
